espresso/authdb/authdb.go

```diff
@@ -30,408 +31,208 @@ func NewAuthDB(db ethdb.Database, mac hash.Hash) (AuthDB, error) {
+func (d *AuthDB) computeMac(key []byte, val []byte) []byte {
 	d.mac.Reset()
-	if err := batch.Put(nextHotshotBlockNumAuthTagKey, tag); err != nil {
-		return fmt.Errorf("failed to put nextHotshotBlockNumAuthTag: %w", err)
-	}
-
-	return nil
+	d.mac.Write(key)
+	d.mac.Write(val)
+	return d.mac.Sum(nil)
 }
 
-func (d *AuthDB) AuthReadNextHotshotBlockNum() (uint64, error) {
-	numBytes, err := d.db.Get(nextHotshotBlockNumKey)
-	if err != nil {
-		if dbutil.IsErrNotFound(err) {
-			return 0, nil
-		}
-		return 0, fmt.Errorf("failed to get nextHotshotBlockNum: %w", err)
-	}
+// either in-memory check (e.g. Freezer) or delegated to `Ancient` (e.g. remotedb.Database),
+// thus safe to pass through
+func (d *AuthDB) HasAncient(kind string, number uint64) (bool, error) {
+	return d.db.HasAncient(kind, number)
+}
 
-	num, err := DecodeUint64(numBytes)
+func (d *AuthDB) Ancient(kind string, number uint64) ([]byte, error) {
+	data, err := d.db.Ancient(kind, number)
 	if err != nil {
-		return 0, fmt.Errorf("failed to decode nextHotshotBlockNum: %w", err)
+		return nil, err
 	}
 
 	if d.mac == nil {
-		return num, nil
+		return data, nil
 	}
 
-	expectedTag, err := d.db.Get(nextHotshotBlockNumAuthTagKey)
-	if err != nil {
-		return 0, fmt.Errorf("failed to get nextHotshotBlockNumAuthTag: %w", err)
-	}
-
-	// verify the auth tag
-	d.mac.Write(nextHotshotBlockNumKey)
-	d.mac.Write(numBytes)
-	tag := d.mac.Sum(nil)
-	d.mac.Reset()
-	if !hmac.Equal(tag, expectedTag) {
-		return 0, fmt.Errorf("failed to verify nextHotshotBlockNumAuthTag for blockNum: %d", num)
-	}
-
-	return num, nil
+	return d.authenticateAncientData(kind, number, data)
 }
 
-// Delayed message fetcher's FromBlock info
-func (d *AuthDB) AuthWriteFromBlock(batch ethdb.Batch, fromBlk uint64) error {
-	if err := batch.Put(fromBlockKey, EncodeUint64(fromBlk)); err != nil {
-		return fmt.Errorf("failed to put delayedMessageFetcherFromBlock: %w", err)
-	}
-	if d.mac == nil {
-		return nil
-	}
-	d.mac.Write(fromBlockKey)
-	d.mac.Write(EncodeUint64(fromBlk))
-	tag := d.mac.Sum(nil)
+// computeMac computes HMAC for kind, number, and data
+func (d *AuthDB) computeMacForAncient(kind string, number uint64, data []byte) []byte {
 	d.mac.Reset()
-	if err := batch.Put(fromBlockAuthTagKey, tag); err != nil {
-		return fmt.Errorf("failed to put delayedMessageFetcherFromBlockAuthTag for fromBlock: %d", fromBlk)
-	}
-
-	return nil
-}
-
-// Delayed message fetcher's FromBlock info
-func (d *AuthDB) AuthReadFromBlock() (uint64, error) {
-	numBytes, err := d.db.Get(fromBlockKey)
-	if err != nil {
-		if dbutil.IsErrNotFound(err) {
-			return 0, nil
+	d.mac.Write([]byte(kind))
+	d.mac.Write(EncodeUint64(number))
+	d.mac.Write(data)
+	return d.mac.Sum(nil)
+}
+
+// authenticateAncientData verifies authentication for ancient data
+func (d *AuthDB) authenticateAncientData(kind string, number uint64, data []byte) ([]byte, error) {
+	switch kind {
+	case rawdb.ChainFreezerHashTable:
+		dataSize := len(data) - d.mac.Size()
```

Comment:
I am not sure I understand what you are doing here:
```
dataSize := len(data) - d.mac.Size()
expectedTag := data[dataSize:]
actualData := data[:dataSize]
```

Why is d.mac.Size() relevant here if we always reset the mac?

Reply:
when we append `ChainFreezerHashTable`, see: 
https://github.com/OffchainLabs/go-ethereum/blob/fcd6eade1b62c94eb144a40fc9a7f0419e9e6674/core/rawdb/accessors_chain.go#L748-L763

 geth use `AppendRaw` and in our auth version of the `AppendRaw`, we append the tag to the raw data bytes, thus the first `len - 32` bytes are the data, the last 32 bytes is the tag in authenticated ancient store.

>Why is d.mac.Size() relevant here if we always reset the mac?

it sounds like you misunderstood what `Hash.Size()` means, for Sha256, it's always 32:
https://cs.opensource.google/go/go/+/master:src/crypto/internal/boring/sha.go;l=259

it's not the accounting of how many bytes it has accumulated so far.